### DISCUSSION TITLE
In custom formulas, how can I get CELL_RANGE, COLUMN_RANGE, etc values?

### DISCUSSION THREAD

**Discussion by nathipg** [handsontable/hyperformula#778] [General]
Hello!

I’m starting to create custom formulas and I’m kind of confused with how to get the values.

By now I have this simple code:

```
min(ast, state) {
  console.log(ast, state);
}
```

Being called like this:

`=MIN(1,2,A1:A6, A:A)`

Which outputs that:

![image](https://user-images.githubusercontent.com/14853904/126503828-d2c38f60-d693-436d-a7a9-95d8a1e23094.png)

My question is: How do I get the value of those ranges? I looked at HyperFormula documentation, but there’s only a sample that doesn’t use the incoming parameters

**Comment by MartinDawson** [comment]
I also have a similar problem which I think may be related to the same as @nathipg issue above. Although I'm not sure it's actually a supported feature of custom functions yet. As far as I can tell there is no way to get these values from non simple args?

I.e our formula looks like this:

`=FIN("revenue", =FIN("date", -1))`

Produces this:

![image](https://user-images.githubusercontent.com/15030491/126627496-4d34300b-d648-4c38-b9b3-57c6534b5856.png)

I would expect the inner formulas to be executed first by hyperformula and passed as the value in through the args param but that doesn't seem to be the case.

**Comment by wojciechczerniak** [comment]
All our plugins and the docs go through `runFunction`, you can check this function body how arguments are coerced: https://github.com/handsontable/hyperformula/blob/413e199a632452bffb45f33c747b53a6f8af2cab/src/interpreter/plugin/FunctionPlugin.ts#L352 or use it to do the coercion for you that will be based on function meta your provide

**Reply by MartinDawson** [reply]
I see, thanks. It's not obvious in the docs that this function can do that.

https://handsontable.github.io/hyperformula/guide/custom-functions.html#implementing-your-custom-function

The example has this:

```
// wrap your custom function in `runFunction()`
  public hyper(ast, state) {
    return this.runFunction(() => 'Hyperformula'.length)
  } 
```
  
I think would be good to cover the above common scenario's as well. It just says it covers validation at the moment but not that you can actually coerce arguments.

**Reply by wojciechczerniak** [reply]
True. Current tutorial has a simple, introductory example. We should think about more advanced guide. CC @kirszenbaum

**Reply by nathipg** [reply]
> All our plugins and the docs go through `runFunction`, you can check this function body how arguments are coerced:
> 
> https://github.com/handsontable/hyperformula/blob/413e199a632452bffb45f33c747b53a6f8af2cab/src/interpreter/plugin/FunctionPlugin.ts#L352
> 
> or use it to do the coercion for you that will be based on function meta your provide

Thanks for reply!
I have a question, how is this done in 0.6.2?
I tried to make a sample but it gives an error

Sample 1 (With one argument):
https://codesandbox.io/embed/priceless-water-z0r3l?fontsize=14&hidenavigation=1&theme=dark

Sample 2 (With three arguments as I saw in source code):
https://codesandbox.io/s/cool-dust-sx8bz?file=/src/countHFFunctionPlugin.js

In 0.6.2 docs, I didn't find an example

**Reply by wojciechczerniak** [reply]
Well, you would have to switch to 0.6.2 tag and go from there:
https://github.com/handsontable/hyperformula/blob/43071aa2a822ddcea04a7b1016fa3e8dda78e751/src/interpreter/plugin/FunctionPlugin.ts#L270-L277

As you can see the logic was in `runFunctionTemplate` function back then. A lot has changed.

**Reply by nathipg** [reply]
Thank you for helping me! I didn't mention here, but I managed to parse the arguments after some discussions with @taknb2nch

**Comment by izulin** [comment]
One thing: function is responsible for calling `Interpreter` on its arguments. So e.g. `this.evaluateAst(ast, state)` for `ast` iterating over its arguments ast should be called.

**Reply by nathipg** [reply]
> One thing: function is responsible for calling `Interpreter` on its arguments. So e.g. `this.evaluateAst(ast, state)` for `ast` iterating over its arguments ast should be called.

Thank you for reply!
Is this for 0.6 or 1.0 version? I tried to make a sample, but a get a "Maximum call stack size exceeded" error

https://codesandbox.io/s/boring-pare-1umzi?file=/src/countHFFunctionPlugin.js

**Reply by taknb2nch** [reply]
I'm in simlar situation.
I set `=HYPER(1, 2, A4:B5, A:A)` and tried the code below.
```
public hyper(ast, formulaAddress) {
  ast.args.forEach(e => {
    let v = this.evaluateAst(e, formulaAddress);

    if (v['valuesFromTopLeftCorner']) {
      let vs = v['valuesFromTopLeftCorner']();

      vs.forEach(e1 => {
        console.info('RANGE', e1);
      });
    } else {
      console.info('NUMBER', v);
    }
  });

  return 'Hyperformula'.length;.
}
```
I think I've got the value I expected (except `Symbol(Empty value)`), but I'm not sure if it's the right way to go.

**Reply by nathipg** [reply]
> I'm in simlar situation.
> I set `=HYPER(1, 2, A4:B5, A:A)` and tried the code below.
> 
> ```
> public hyper(ast, formulaAddress) {
>   ast.args.forEach(e => {
>     let v = this.evaluateAst(e, formulaAddress);
> 
>     if (v['valuesFromTopLeftCorner']) {
>       let vs = v['valuesFromTopLeftCorner']();
> 
>       vs.forEach(e1 => {
>         console.info('RANGE', e1);
>       });
>     } else {
>       console.info('NUMBER', v);
>     }
>   });
> 
>   return 'Hyperformula'.length;.
> }
> ```
> 
> I think I've got the value I expected (except `Symbol(Empty value)`), but I'm not sure if it's the right way to go.

Thank you very much!
Not sure if you had this issue, but If you need to check if cell is empty, you can import EmptyValue from 'hyperformula/es/interpreter/InterpreterValue', not sure if is the best way, but I'm doing it

**Reply by taknb2nch** [reply]
Thank you for your suggestion.

I changed to import to `hyperformula/es/...` and tried to implement it like existing plugin (I read MedianPlugin).

```
import { FunctionPlugin, ErrorType } from 'hyperformula';
import { RawScalarValue } from 'hyperformula/typings/interpreter/InterpreterValue'
import { CellError } from 'hyperformula/es/Cell'
import { ErrorMessage } from 'hyperformula/es/error-message'
import { ArgumentTypes } from 'hyperformula/es/interpreter/plugin/FunctionPlugin';

export class CustomFunctions extends FunctionPlugin {
  public static implementedFunctions = {
    'HYPER': {
      method: 'hyper',
      expandRanges: true,
      parameters: [
        {argumentType: ArgumentTypes.ANY},
      ],
      repeatLastArgs: 1,
    }
  };

  public hyper(ast, formulaAddress) {
    return this.runFunction(ast.args, formulaAddress, this.metadata('HYPER'), (...args: RawScalarValue[]) => {
      const values = this.interpreter.arithmeticHelper.coerceNumbersExactRanges(args);

      if(values instanceof CellError) {
        return values;
      }

      const vs = values as number[];

      if (vs.length === 0) {
        return new CellError(ErrorType.NUM, ErrorMessage.OneValue)
      }

      console.info(vs);
      
      return 'Hyperformula'.length;
    });
  }
}
```
It seems to be working fine.

I'm not very familiar with TypeScript.
```
import { CellError } from 'hyperformula/typings/Cell';
```
get a build error, but 
```
import { CellError } from 'hyperformula/es/Cell';
```
dose not.

I don't know why.
Please let me know.

**Reply by nathipg** [reply]
I gave a look at the files, but I couldn't find a reason to not work properly. I'm new with typescript either, so I might be missing something.
Not sure why, but my VSCode does not autocomplete to "typings", only to "es", thats why I've imported the other file right away.

**Reply by wojciechczerniak** [reply]
CellError is exported as `impot { CellError } from 'hyperformula'` and Typescript should choose /es dir automatically based on `package.json`. Also, types and autocomplete will use definition file that is generated for those named exports.

Full list you can see here: https://github.com/handsontable/hyperformula/blob/413e199a632452bffb45f33c747b53a6f8af2cab/src/index.ts#L113-L173

If something is missing pls let us know

**Reply by nathipg** [reply]
Thank you! I'll update my imports

**Comment by sequba** [comment]
Hi @nathipg, @taknb2nch 

I'm more than happy to announce that we just released a new [custom functions guide](https://hyperformula.handsontable.com/guide/custom-functions.html) with an advanced example. You should check it out ;)

**Reply by taknb2nch** [reply]
Thank you @sequba.
I'll check it out.
